src/NuGet.Core/NuGet.ProjectModel/TargetFrameworkInformation.cs

```diff
@@ -143,9 +135,170 @@ public bool Equals(TargetFrameworkInformation other)
                    StringComparer.OrdinalIgnoreCase.Equals(TargetAlias, other.TargetAlias);
         }
 
-        public TargetFrameworkInformation Clone()
+        /// <summary>
+        /// Returns copy of this with specified versions defined in the Central package versions management file. 
+        /// </summary>
+        public TargetFrameworkInformation WithCentralPackageVersions(ImmutableDictionary<string, CentralPackageVersion> versions)
+        {
+            versions ??= EmptyCentralPackageVersions;
+
+            if (CentralPackageVersions == versions)
+            {
+                return this;
+            }
+
+            return new TargetFrameworkInformation(this)
+            {
+                CentralPackageVersions = versions
+            };
+        }
+
+        /// <summary>
+        /// Returns copy of this with specified framework references. 
+        /// </summary>
+        public TargetFrameworkInformation WithFrameworkReferences(ImmutableHashSet<FrameworkDependency> references)
         {
-            return new TargetFrameworkInformation(this);
+            references ??= ImmutableHashSet<FrameworkDependency>.Empty;
+
+            if (FrameworkReferences == references)
+            {
+                return this;
+            }
+
+            return new TargetFrameworkInformation(this)
+            {
+                FrameworkReferences = references
+            };
+        }
+
+        /// <summary>
+        /// Returns copy of this with specified download dependencies 
+        /// </summary>
+        public TargetFrameworkInformation WithDownloadDependencies(ImmutableArray<DownloadDependency> dependencies)
+        {
+            if (DownloadDependencies.Equals(dependencies))
+            {
+                return this;
+            }
+
+            return new TargetFrameworkInformation(this)
+            {
+                DownloadDependencies = dependencies
+            };
+        }
+
+        /// <summary>
+        /// Returns copy of this with specified asset target fallback
+        /// </summary>
+        public TargetFrameworkInformation WithAssetTargetFallback(bool assetTargetFallback)
+        {
+            if (AssetTargetFallback == assetTargetFallback)
+            {
+                return this;
+            }
+
+            return new TargetFrameworkInformation(this)
+            {
+                AssetTargetFallback = assetTargetFallback
+            };
+        }
+
+        /// <summary>
+        /// Returns copy of this with specified warn
+        /// </summary>
+        public TargetFrameworkInformation WithWarn(bool warn)
```

Comment:
I didn't go through the ones below this method, but I think similar logic applies.